tiledb/sm/consolidator/fragment_consolidator.cc

```diff
@@ -407,6 +407,60 @@ Status FragmentConsolidator::consolidate_fragments(
         std::to_string(fragment_uris.size()) + " required fragments.");
   }
 
+  // In case we have a dense Array check that the fragments can be consolidated
+  // without data loss. More specifically, if the union of the non-empty domains
+  // of the fragments which are selected for consolidation (which is equal to
+  // the non-empty domain of the resulting consolidated fragment) overlaps with
+  // any fragment created prior to this subset, then the subset is marked as
+  // non-consolidatable. Recall that the fragment that results from
+  // consolidating a subset of fragments containing at least one dense fragment
+  // is always a dense fragment. Therefore, empty regions in the non-emtpy
+  // domain of the consolidated fragment will be filled with special values.
+  // Those values may erroneously overwrite older valid cell values
+  if (array_for_reads->array_schema_latest().array_type() == ArrayType::DENSE) {
+    // Search every other fragment in this array if any of them overlaps in
+    // ranges and its timestamp range falls between the range of the fragments
+    // to consolidate throw error
+
+    // First calculate the timestamp bounds
+    uint64_t max_timestamp = std::numeric_limits<uint64_t>::min();
+    for (const auto& item : to_consolidate) {
+      const auto& range = item.timestamp_range();
+      max_timestamp = std::max(max_timestamp, range.second);
+    }
+
+    // Now iterate all fragments and see if the consolidation can lead to data
+    // loss
+    for (auto& frag_info : frag_info_vec) {
+      // Ignore the fragments that are requested to be consolidated
+      auto uri = frag_info.uri().last_path_part();
+      if (to_consolidate_set.count(uri) != 0) {
+        continue;
+      }
+
+      // Expand domain to full tiles
+      auto expanded_non_empty_domain = union_non_empty_domains;
+      domain.expand_to_tiles(&expanded_non_empty_domain);
+
+      // Check domain and timestamp overlap
```

Comment:
We should do timestamp check first as it's cheaper. We should also document that we use start timestamp.

Reply:
Done!